src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java

```diff
@@ -0,0 +1,173 @@
+        return new J2EContext(request, response);
+    }
+
+
+    /**
+     * @return a SAML2Client object to interact with the IdP service.
+     */
+    protected SAML2Client createSAML2Client() {
+        //FIXME [kuisathaverat] Throws NPE I do not like it
+        Preconditions.checkNotNull(samlPluginConfig.getIdpMetadata());
+
+        final SAML2ClientConfiguration config = new SAML2ClientConfiguration();
+        config.setIdentityProviderMetadataResource(new RewriteableStringResource(samlPluginConfig.getIdpMetadata()));
+        config.setDestinationBindingType(SAML2_REDIRECT_BINDING_URI);
+
+        /* TODO [kuisathaverat] add settings to configure algorithms
+        config.setBlackListedSignatureSigningAlgorithms(...);
+        config.setSignatureAlgorithms(...);
+        config.setSignatureReferenceDigestMethods(...);
+        config.setSignatureCanonicalizationAlgorithm(...);
+        */
+
+        if (samlPluginConfig.getEncryptionData() != null) {
+            config.setWantsAssertionsSigned(true);
+            config.setKeystorePath(samlPluginConfig.getEncryptionData().getKeystorePath());
+            config.setKeystorePassword(samlPluginConfig.getEncryptionData().getKeystorePassword());
+            config.setPrivateKeyPassword(samlPluginConfig.getEncryptionData().getPrivateKeyPassword());
+        } else {
+            if (!KS.isValid()) {
+                KS.init();
+            }
+            if (KS.isUsingDemoKeyStore()) {
+                LOG.warning("Using bundled keystore is INSECURE: " + KS.getKeystorePath());
+            }
+            config.setKeystorePath(KS.getKeystorePath());
+            config.setKeystorePassword(KS.getKsPassword());
+            config.setPrivateKeyPassword(KS.getKsPkPassword());
+        }
+
+        config.setMaximumAuthenticationLifetime(samlPluginConfig.getMaximumAuthenticationLifetime());
+
+        if (samlPluginConfig.getAdvancedConfiguration() != null) {
+
+            // request forced authentication at the IdP, if selected
+            config.setForceAuth(samlPluginConfig.getForceAuthn());
+
+            // override the default EntityId for this SP, if one is set
+            if (samlPluginConfig.getSpEntityId() != null) {
+                config.setServiceProviderEntityId(samlPluginConfig.getSpEntityId());
+            }
+
+            // if a specific authentication type (authentication context class
+            // reference) is set, include it in the request to the IdP, and request
+            // that the IdP uses exact matching for authentication types
+            if (samlPluginConfig.getAuthnContextClassRef() != null) {
+                config.setAuthnContextClassRef(samlPluginConfig.getAuthnContextClassRef());
+                config.setComparisonType("exact");
+            }
+        }
+
+        final SAML2Client saml2Client = new SAML2Client(config);
+        saml2Client.setCallbackUrl(samlPluginConfig.getConsumerServiceUrl());
```

Comment:
What's this configuration for? AFAIK the callback URL is set on the IdP (linked to a service provider ID).